egui/src/widgets/plot/legend.rs

```diff
@@ -1,81 +1,231 @@
-use std::string::String;
+use std::{
+    collections::{BTreeMap, HashSet},
+    string::String,
+};
 
 use crate::*;
 
-pub(crate) struct LegendEntry {
-    pub text: String,
-    pub color: Color32,
-    pub checked: bool,
-    pub hovered: bool,
+use super::items::PlotItem;
+
+/// Where to place the plot legend.
+#[derive(Debug, Clone, Copy, PartialEq)]
+pub enum LegendPosition {
+    TopLeft,
+    TopRight,
+    BottomLeft,
+    BottomRight,
+}
+
+impl LegendPosition {
+    pub fn all() -> impl Iterator<Item = LegendPosition> {
+        [
+            LegendPosition::TopLeft,
+            LegendPosition::TopRight,
+            LegendPosition::BottomLeft,
+            LegendPosition::BottomRight,
+        ]
+        .iter()
+        .copied()
+    }
+}
+
+/// The configuration for a plot legend.
+#[derive(Clone, Copy, PartialEq)]
+pub struct Legend {
+    pub text_style: TextStyle,
+    pub position: LegendPosition,
+}
+
+impl Default for Legend {
+    fn default() -> Self {
+        Self {
+            text_style: TextStyle::Body,
+            position: LegendPosition::TopRight,
+        }
+    }
+}
+
+#[derive(Clone)]
+struct LegendEntry {
+    color: Color32,
+    checked: bool,
+    hovered: bool,
 }
 
 impl LegendEntry {
-    pub fn new(text: String, color: Color32, checked: bool) -> Self {
+    fn new(color: Color32, checked: bool) -> Self {
         Self {
-            text,
             color,
             checked,
             hovered: false,
         }
     }
 }
 
-impl Widget for &mut LegendEntry {
+impl Widget for (&String, &mut LegendEntry) {
```

Comment:
This is a bit weird. Maybe just make a function for it:

``` rust
fn legend_ui(ui: &mut Ui, text: &str, legend: &mut LegendEntry) {
…
}
```

?

Reply:
Made it a method of `LegendEntry.